keps/sig-network/3726-standard-application-protocols/README.md

```diff
@@ -0,0 +1,236 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There are cases where implementations implement different things for the same application protocol names. That already causes(and can cause more in the future) issues for certain implementations to interoperate. (see example with GKE and Istio under [Motivation](#motivation))
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA service names.
+
+
+## Motivation
+
+The lack of direct support for specifying application protocols for ports and the widespread use of implementation-specific annotations to mitigate it had led k8s to add the `AppProtocol` field to the port spec.
+
+While this is a good solution - we never came with recommended standards other than [IANA standard service names](https://www.iana.org/assignments/service-names) or a domain prefixed protocol.
+
+This loose definition has led us to have instances where implementations do different things for common protocols.
+
+One good example is `HTTP2`. 
+* In GKE you can use `appProtocol: HTTP2` and it will describe HTTP2 over TLS (https://cloud.google.com/kubernetes-engine/docs/how-to/secure-gateway#load-balancer-tls).
+* While in Istio it will be h2c (HTTP2 over cleartext).
+
+That creates a problem where users with GKE and Istio in their cluster can have very different behaviors for the same `appProtocol` value.
+
+
+
+### Goals
+
+* Build consensus around how common (non IANA service names) should be implemented
+* Help the broader community and specifically implementations to interoperate better
+* Provide short and clear documentation for how appProtocol values should be implemented
+  * Update the appProtocol user documentation respectively
+
+
+### Non-Goals
+
+* Validate appProtocol values
+* Monitor `appProtocol implementations
+
+
+## Proposal
+
+Kubernetes supports the `appProtocol` field to provide a way to specify an application protocol for each Service port.
+
+The [documentation](https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol) for this field says that:
+
+```Values should either be IANA standard service names or domain prefixed names such as mycompany.com/my-custom-protocol.```
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA standard service names.
```

Comment:
where is going to live this list?

Reply:
I think that the KEP should answer the where, who and how this list is going to be created, maintained, ... what are the requirements ... or we can be bikeshedding for years for each protocol

Reply:
I think the list is small enough that it could live in the spec for AppProtocol, but open to alternatives

Reply:
Do you keep the list versioned matching Kubernetes releases? i.e. the list in 1.26 only applies to 1.26 or the previous one too?

Reply:
Yeah I don't think there's any reason this needs to be versioned, but I also can't think of any other place this could fit. Agree that the KEP needs to define the requirements for protocols that get this status.

Reply:
If I understand correctly(please correct me if not), you are suggesting to have a new repo with standard names and/or structured metadata in it for various things in k8s.

While this can be a good idea, how will we use it for this case? 
We cannot validate the `AppProtocol` field name against it because we are not aware of whether the provided string for this field is an IANA service name, custom impl-specific prefixed name OR something unsupported.

maybe we can reference to the relevant parts of this registry from the field description or from the docs though.

Reply:
are we diligent enough to maintain another repo? 
how do we "enforce" devs to register everything in the new repo?
it looks to me that the idea is good but it got outdated soon, I can see how docs and release folks do a lot of work chasing people to update things.
I like the idea of being in-tree and use some of your "makefile" magic @thockin , that also will allow to be in ` importable Go form`

Reply:
I'll slap together a basic KEP

Reply:
https://github.com/kubernetes/enhancements/pull/3787